Set libnetwork sandbox key w/o OCI hooks

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Made another reexec go away while at the same time [potentially improving compatibility with some OCI runtimes.](https://github.com/kata-containers/kata-containers/issues/5321#issuecomment-1274294450)

**- How I did it**
I made it possible for consumers of libcontainerd to create tasks without immediately starting the user process so that they can run arbitrary code before the process starts. Anything that could be done with a `prestart` or `createProcess` OCI hook can now also be implemented without one. I then replaced the `libnetwork-setkey` OCI hook with an in-daemon equivalent.

**- How to verify it**
CI

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
- OCI lifecycle hooks are no longer used to configure container networking. This may improve the compatibility with some alternative runtimes.

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
daemon: set libnetwork sandbox key w/o OCI hook

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libnetwork: remove unused libnetwork-setkey reexec

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
builder-next: setup network namespace w/o OCI hook

Have libnetwork create the network namespace and pass the path to the
namespace to runC. Switch to buildkit's containerd executor when
userns-remapping is enabled so the in-process OnCreateRuntime callback
can be used in place of the OCI hook.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented:
I realized that builder-next can utilize user namespaces via the `IdentityMapping` option passed into `runcexecutor`, therefore my change to builder-next breaks buildkit builds when the daemon is started with the `--userns-remap` option. (There must be a gap in integration-test coverage as CI didn't catch that breakage.) That leaves us with a few options:
1. Buildkit's `runcexecutor` grows a callback function which is called between the create and start operations: an in-process analogue to the `createProcess` OCI hook
2. We keep the `libnetwork-setkey` reexec hook around in some form or another
3. Park this PR until more of the containerd snapshotter integration has been merged, as the builder-next executor is being switched over to use the buildkit's `containerdexecutor` as part of that effort. That executor would still need to grow an analogous callback to the `createProcess` OCI hook, but it would be a less-invasive change than on `runcexecutor` due to the containerd client forcing the separation between `NewTask` (i.e. create) and `Start` operations.

tonistiigi commented:
I wonder if we should have a separate codepath for userns and non-userns in this case. Eventually I'd like to have the same netns pooling in buildkit in dockerd that is in upstream as it is the most performant but based on your description that does not seem possible. I assume that dockerd making the user namespace fd itself as well and passing it to runc(then I would assume it can create netns associated with same userns) isn't an option either?

corhere commented:
Multithreaded processes cannot change their user namespace. The kernel will refuse to setns or unshare the user namespace of any task which shares its virtual memory space with any other task. So while dockerd technically _could_ make the user namespace fd itself, it would have to fork a whole new process to do so. I don't see any advantages over deferring user-namespace creation to the runtime. We could pool and reuse runtime-created user and network namespaces if we wanted to; the daemon merely has to persist them through the usual mechanisms (holding open an fd, bind-mounting) before the container is stopped.

corhere committed:
builder-next: pass whole Opt struct to newExecutor

The options required by the executor depend on the platform, and soon
will also depend on the values of other options. Give the executor
constructor the flexibility to pull whatever options it needs out of the
Opts struct.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented:
I made progress towards #44690, only to discover that `containerdexecutor` does not have user-namespace support plumbed in. More work is needed on the buildkit side.

corhere commented:
Unfortunately, runc invokes the prestart OCI hooks before it applies the sysctls in the container spec, contrary to what the OCI runtime spec says MUST be done. Moving setting the libnetwork sandbox key to after the create operation completes is therefore a breaking change.
